### PR TITLE
Ast to cfg

### PR DESCRIPTION
**Pull request by jfet97**
Under inspection

### CODE REVIEW

File: lib/cfg/cfg_Base.ml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+module CFG = struct
+  (* Auxiliary module for providing Hashtbl of a pretty printing function *)
+  module Hashtbl = struct
+    include Hashtbl
+
+    let pp pp_key pp_value ppf values =
+      Hashtbl.iter (fun key data -> Format.fprintf ppf "@[<1>%a: %a@]@." pp_key key pp_value data) values
+  end
+
+  type 'a item = {
```
Comment (jfet97) [inline lib/cfg/cfg_Base.ml:13]: please add a brief description about what an item is and a short description about its 4 fields

File: lib/cfg/cfg_Converter.mli [hunk 2]
```diff
@@ -0,0 +1,10 @@
+open Ast.HeapRegularCommands
+open Cfg_Node
+
+(** The module Converter provides a method [convert], that given an AST, it converts it
+   into a CFG composed by nodes from the {{! Cfg.Node}Node} module.
+   See {{! Cfg.CFG.make}make} to convert the resulting structure to a CFG.
+ *)
+module Converter : sig
+  val convert : ?keep_structure:bool -> 'a HeapRegularCommand.t -> 'a HeapAtomicCommand.t list Node.t
```
Comment (jfet97) [inline lib/cfg/cfg_Converter.mli:12]: A comment about the purpose of the `keep_structure` flag would be appreciated here

File: lib/cfg/cfg_Base.ml [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+  let idx (_ : 'a t) (item: 'a item) =
+    item.idx
+
+  (* Auxiliary module used into the fold function *)
+  module IntSet = Set.Make(struct
+    type t = int
+    let compare = compare
+  end)
+
+  let fold (cfg : 'a t) (fn: 'a t -> 'a item -> 'b -> 'b) (acc: 'b) =
```
Comment (jfet97) [inline lib/cfg/cfg_Base.ml]: Why don't you filter out already visited successors before calling visit recursively instead of wasting recursive calls on them?

File: lib/cfg/cfg_Base.ml [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  end
+
+  type 'a item = {
+    idx   : int;
+    exp   : 'a;
+    pred  : int list;
+    succ  : int list;
+  } [@@deriving show]
+
+  (* The CFG is implemented as an Hashtable <id, (exp, predecessors, successors)> *)
```
Comment (jfet97) [inline lib/cfg/cfg_Base.ml]: If I'm understanding the code correctly, this `id` is the same as the `idx` field of the following `item` definition, so why the two things have different names? This is confusing

The comment should be: (* The CFG is implemented as an Hashtable <idx, (idx, exp, predecessors, successors)> *)

File: lib/cfg/cfg_Base.ml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    {idx = idx; exp = exp; pred = pred; succ = succ}
+
+  (* starting from a node (the root), builds a CFG *)
+  let make (initial_node : 'a Node.t) : 'a t =
+    let h = Hashtbl.create (Node.length initial_node) in
+
+    let rec aux (h : (int, 'a item) Hashtbl.t) (node : 'a Node.t) : unit =
+      match Node.get_succ node with
+      | [] -> (* the node has not successors, insert it *)
+         Hashtbl.add h (Node.get_id node) (make_item (Node.get_id node) (Node.get_exp node) (Node.get_pred node) [])
```
Comment (jfet97) [inline lib/cfg/cfg_Base.ml]: Could you please extract this in a separate function? To then use the new function also in the code below

File: lib/cfg/cfg_Base.mli [hunk 6]
```diff
@@ -1,9 +1,15 @@
+open Cfg_Node
```
Comment (jfet97) [inline lib/cfg/cfg_Base.mli:1]: A comment for the whole file, in few lines you call the same things in three different ways: index, idx and id. This is confusing

File: lib/cfg/cfg_Node.mli [hunk 7]
```diff
@@ -0,0 +1,54 @@
+(** The module provides the abstraction of a CFG node. *)
+module Node : sig
+  type 'a t [@@deriving show, sexp]
+
+  (** Checks if two nodes and their children are the same recursively.
+
+      Notes: Should always work for the structure generated by {{! Cfg.Converter}Converter}.
+      Severely limited, only checks isomorphism for two children for each node,
```
Comment (jfet97) [inline lib/cfg/cfg_Node.mli:8]: Why there are these limitations? What are the possible problems we should be aware of?

File: lib/cfg/cfg_Node.mli [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  (** Replace the metadata of the node *)
+  val set_exp : 'a t -> 'a -> unit
+
+  (** Replace the succ list of the node *)
+  val set_succ : 'a t -> 'a t list -> unit
+
+  (** Modifies the succ lists such that no loops are present, destroyes
+      the original topology, pred lists are not modified *)
+  val structure_without_loops_destructive : 'a t -> unit
```
Comment (jfet97) [inline lib/cfg/cfg_Node.mli:53]: Why is this necessary?

File: lib/cfg/cfg_Node.ml [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+  !counter
+
+module Node = struct
+  (* Recursive definition of CFG node, made by
+  *  id,
+  *  exp,
+  *  successors,
+  *  predecessors
+  *)
+  type 'a t = {
```
Comment (jfet97) [inline lib/cfg/cfg_Node.ml:16]: Why is this duplicated? Why isn't the `item` structure in `cfg_Base` enough?

File: lib/cfg/cfg_Node.mli [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  (** [make data succ prec] creates a new unique node using [data] as its content,
+    [succ] the list of successors and [prec] the list of precedessors identifiers. *)
+  val make : 'a -> 'a t list -> int list -> 'a t
+
+  (** [make id data succ prec] creates a new node using [id] as its identifiers, [data] as its content,
+    [succ] the list of successors and [prec] the list of precedessors identifiers.
+    
+    *Warning* if the same id as another node is chosen, other methods might return wrong results.
+   *)
+  val make_with_id : int -> 'a -> 'a t list -> int list -> 'a t
```
Comment (jfet97) [inline lib/cfg/cfg_Node.mli:25]: Why is the possibility of setting a custom id exposed?

File: lib/cfg/cfg_Node.mli (hunk 7 above)
Comment (jfet97) [reply lib/cfg/cfg_Node.mli:8]: And why we have to compare two nodes?

File: lib/cfg/cfg_Converter.ml [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+       Node.add_succ dummyentry root2;
+       Node.add_succ !end1 dummyexit;
+       Node.add_succ !end2 dummyexit;
+       (dummyentry, ref dummyexit)
+    | Star(comm) ->
+       let (root1, end1) = convert_helper comm in
+       Node.add_succ !end1 root1;
+       (root1, end1)
+
+  let simplify (keep_structure: bool) (root: 'a HeapAtomicCommand.t list Node.t) : unit =
```
Comment (jfet97) [inline lib/cfg/cfg_Converter.ml:45]: please add a comment to explain the overall behaviour of this function, why we need it, what it does, who are we, ma quando si mangia?

File: lib/cfg/cfg_Converter.ml [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+        )
+        | [a] when Node.get_id a = Node.get_id root -> ()
+        | a -> List.iter (help_simplify keep_structure) a
+      )
+    in
+    help_simplify keep_structure root
+
+  let convert ?(keep_structure: bool = true) (root: 'a HeapRegularCommand.t) : 'a HeapAtomicCommand.t list Node.t =
+    match convert_helper(root) with
+    | (a, _) -> simplify keep_structure a; a
```
Comment (jfet97) [inline lib/cfg/cfg_Converter.ml]: "non mi piace" -Ale

File: lib/cfg/cfg_Converter.ml [hunk 13]
```diff
@@ -0,0 +1,79 @@
+open Ast.HeapRegularCommands
+open Cfg_Node
+
+module Converter = struct
+  let rec convert_helper(root: 'a HeapRegularCommand.t) :
```
Comment (jfet97) [inline lib/cfg/cfg_Converter.ml:10]: Why is a reference to the same node needed?

File: lib/cfg/cfg_Converter.ml [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+       (dummyentry, ref dummyexit)
+    | Star(comm) ->
+       let (root1, end1) = convert_helper comm in
+       Node.add_succ !end1 root1;
+       (root1, end1)
+
+  let simplify (keep_structure: bool) (root: 'a HeapAtomicCommand.t list Node.t) : unit =
+    let alreadyvisited = ref [] in
+
+    let rec apply_and_recurse (root) (a) (keep_structure) : unit =
```
Comment (jfet97) [inline lib/cfg/cfg_Converter.ml]: please use a more descriptive name than `a`

Comment (jfet97) [reply lib/cfg/cfg_Converter.ml]: There is an implicit assumption for the correctness: `a` is the only successor of `root`

File: lib/cfg/cfg_Node.mli (hunk 7 above)
Comment (elvilio) [reply lib/cfg/cfg_Node.mli:8]: The limitations are because graph ismorphism is not feasable, the problems are simply that it may not return true if the two nodes are actually the same.
We need to compare two nodes for the unit tests, otherwise it's not really needed.

File: lib/cfg/cfg_Node.mli (hunk 8 above)
Comment (elvilio) [reply lib/cfg/cfg_Node.mli:53]: Because printing a graph structure is probably not in the scope of `[@@deriving show]` since it doesn't terminate.

File: lib/cfg/cfg_Node.mli (hunk 10 above)
Comment (elvilio) [reply lib/cfg/cfg_Node.mli:25]: Because to create a looping structure manually is otherwise impossible unless you provide the AST and convert, but in the tests we need the manual approach.

File: lib/cfg/cfg_Converter.ml (hunk 14 above)
Comment (elvilio) [reply lib/cfg/cfg_Converter.ml]: > There is an implicit assumption for the correctness: `a` is the only successor of `root`

Because it is.

File: lib/cfg/cfg_Converter.ml (hunk 12 above)
Comment (elvilio) [reply lib/cfg/cfg_Converter.ml]: Cosa non ti piace?

File: lib/cfg/cfg_Converter.ml (hunk 13 above)
Comment (elvilio) [reply lib/cfg/cfg_Converter.ml:10]: Where?

Comment (elvilio) [reply lib/cfg/cfg_Converter.ml:10]: Because we don't need to pass the whole CFG around, it might be expensive. We need to keep in memory the last node of the CFG otherwise we can't grow the structure.

**Comment by Yurand2000** [comment]
Already merged in the branch Analysis, all the code will be reviewed there before merging into the main branch.
